src/nvme/tree.c

```diff
@@ -1959,34 +2013,37 @@ static void nvme_ns_set_generic_name(struct nvme_ns *n, const char *name)
 static nvme_ns_t nvme_ns_open(const char *name)
 {
 	struct nvme_ns *n;
+	int fd;
 
 	n = calloc(1, sizeof(*n));
 	if (!n) {
 		errno = ENOMEM;
 		return NULL;
 	}
 
+	n->fd = -1;
 	n->name = strdup(name);
-	n->fd = nvme_open(n->name);
-	if (n->fd < 0)
+
+	fd = nvme_ns_get_fd(n);
+	if (fd < 0)
 		goto free_ns;
 
 	nvme_ns_set_generic_name(n, name);
 
-	if (nvme_get_nsid(n->fd, &n->nsid) < 0)
-		goto close_fd;
+	if (nvme_get_nsid(fd, &n->nsid) < 0)
+		goto free_ns;
 
 	if (nvme_ns_init(n) != 0)
-		goto close_fd;
+		goto free_ns;
 
 	list_head_init(&n->paths);
 	list_node_init(&n->entry);
 
+	nvme_ns_release_fd(n); /* Do not leak fds */
```

Comment:
Alright, so the aim is to get rid of the fd eventually so we don't need to open it at all.

Reply:
That was actually not what I had in mind.
`# cat /sys/block/nvme0n1/nsid`
provides you with the same result; we should rather read _all_ information from sysfs and _never_ open a filedescriptor for the initial scanning. Doing so will block if the namespace/controller is undergoing a reset, and hence 'nvme disconnect' will hang for systems where the controller is at fault.

Reply:
Yes, this where we aiming towards. This is towards this goal.

Reply:
Personally, I would prefer to never cache the fds and always open/close before/after every read/write/ioctl operations. But that would mean a big change in the way libnvme works and would require users to always remember to close the fd after use. 

Like Daniel said, this PR is a step towards better handling of file descriptors.

@igaw - I will make the changes you suggested this morning.